### PR TITLE
Functor: Add functionality so that Functor can contain a free function

### PR DESCRIPTION
**Pull request by kwikius**
( or static member function) as well as an object pointer and a non-static member function. The size and performance of the functor is not affected. A list of functors , for example, can freely mix, object member functions, static member functions and free functions.

A new constructor is added, that takes a pointer to free function argument.
One consequence, when explicitly initialising an empty functor, you need to be explicit about constructing with a nullptr. Constructing with a 0 or NULL wont work since that is convertible to a function pointer as well as a nullptr

There is also an overload of the bind static member for the free function pointer argument

### CODE REVIEW

**Comment by dgrat** [comment]
@kwikius 
I would probably try to replace the macros with templates. I am more or less sure this is definitely possible. Keep in mind that macros ignore namespaces.

**Comment by kwikius** [comment]
@dgrat.

Funny. I had started doing that :)
https://github.com/kwikius/ardupilot/blob/functor_mod/libraries/AP_Scheduler/AP_Scheduler.h#L58
https://github.com/kwikius/ardupilot/blob/functor_mod/libraries/AP_Scheduler/examples/Scheduler_test/Scheduler_test.cpp#L45

 It makes the code easier to understand I think, especially in the example which will help people learning the code.

This also has a bearing on https://github.com/ArduPilot/ardupilot/pull/4795. It could be used with it. You may add a bit of overhead but you gain flexibility, because you can assign free functions and static member functions. Ultimately it would be nice to be able to add/remove functions dynamically. This especially might be applicable to  rangefinder, trigger, mount, optical flow, terrain, adsb, logging etc

File: APMrover2/GCS_Mavlink.cpp [hunk 1]
```diff
@@ -1602,9 +1602,9 @@ void Rover::gcs_update(void)
     for (uint8_t i=0; i<num_gcs; i++) {
         if (gcs[i].initialised) {
 #if CLI_ENABLED == ENABLED
-            gcs[i].update(g.cli_enabled == 1 ? FUNCTOR_BIND_MEMBER(&Rover::run_cli, void, AP_HAL::UARTDriver *) : NULL);
+            gcs[i].update(g.cli_enabled == 1 ? FUNCTOR_BIND_MEMBER(&Rover::run_cli, void, AP_HAL::UARTDriver *) : nullptr);
```
Comment (lucasdemarchi) [inline APMrover2/GCS_Mavlink.cpp]: great. but this commit needs to be split across the different components/directories.

File: libraries/AP_Scheduler/AP_Scheduler.h [hunk 2]
```diff
@@ -55,7 +55,7 @@ class AP_Scheduler
     // constructor
     AP_Scheduler(void);
     
-    FUNCTOR_TYPEDEF(task_fn_t, void);
```
Comment (lucasdemarchi) [inline libraries/AP_Scheduler/AP_Scheduler.h]: add "AP_Scheduler:" to the commit message.

File: libraries/AP_HAL/utility/functor.h [hunk 3]
```diff
@@ -87,6 +75,19 @@ class Functor
     }
 
 private:
+
```
Comment (lucasdemarchi) [inline libraries/AP_HAL/utility/functor.h]: remove blank line

File: libraries/AP_HAL/utility/functor.h [hunk 4]
```diff
@@ -87,6 +75,19 @@ class Functor
...
+        , _method(method)
+    {
+    }
+
+    // construct from a free function
+    constexpr Functor( RetType(* pfn)(Args...) )
+        : _obj{reinterpret_cast<void*>(pfn)}
+        , _method{free_wrapper}
+    {
+    }
```
Comment (lucasdemarchi) [inline libraries/AP_HAL/utility/functor.h]: add missing blank line below

**Comment by kwikius** [comment]
Thanks for the comments. Apologies for not getting further on this. I am trying to run the example on px4 with as yet no joy !

**Comment by dgrat** [comment]
In case it may sound rude, it isn't meant like this. I like the concept that you try to improve the functor class, but not how it is now. Atm, you don't buy anything. This is because of the following things:

1) You still use the same macros, the only difference is that you apply std::wrappers. This is a bad idea, because it is not state of the art programming style, it still ignores namespaces and in case of embedded programming it creates a wrapper for every task, which increases the memory consumption. At the same time you don't really know what happens inside of the compile-time generated wrapper, as the standard library is implementation dependent.  

    #define FUNCTOR_TYPEDEF(name, rettype, ...) \
        typedef Functor<rettype, ## __VA_ARGS__> name

    #define FUNCTOR_DECLARE(name, rettype, ...) \
        Functor<rettype, ## __VA_ARGS__> name

    #define FUNCTOR_BIND(obj, func, rettype, ...) \
        Functor<rettype, ## __VA_ARGS__>::bind<std::remove_reference<decltype(*obj)>::type, func>(obj)

    #define FUNCTOR_BIND_MEMBER(func, rettype, ...) \
        Functor<rettype, ## __VA_ARGS__>::bind<std::remove_reference<decltype(*this)>::type, func>(this)

    #define SCHED_TASK_CLASS(classname, classptr, func, _rate_hz, _max_time_micros) { \
        .function = FUNCTOR_BIND(classptr, &classname::func, void),\
        AP_SCHEDULER_NAME_INITIALIZER(func)\
        .rate_hz = _rate_hz,\
        .max_time_micros = _max_time_micros\
    }

2) You are probably also not happy with how it looks, am I right? 

    // here we install a global free function
    {.function = AP_Scheduler::task_fn_t::bind(one_hz_print),
     .name = "one_hz_print",
     .rate_hz = 0.2f,
     .max_time_micros = 1000
    },

Again, it is just my opinion and I don't decide anything here. I noticed already, that coding conventions are pretty bendable in AP. However, it became already better in comparison to what it was before :)

**Comment by kwikius** [comment]
@dgrat Thanks for the comments,but I made no changes to the macros.
I'm closing this anyway since I cant test it on px4 boards for whatever reason, so I am going to work on that

**Comment by dgrat** [comment]
@kwikius  
I think, the macrso would need a change

**Comment by kwikius** [comment]
@dgrat  @lucasdemarchi  Enough FUD thankyou regarding arm gcc. I suggest you substantiate your claims by testing before coming out with these daft comments

**Comment by dgrat** [comment]
@kwikius 
This is why I asked for an illustration. .

**Comment by kwikius** [comment]
So far we had std::function fails to compile on arm-gcc   and now std::remove_reference increases memory consumption on arm-gcc.  The first is wrong,  The second one is just hysterical :) I say again, I made no changes to those macros. They were already there.

**Comment by dgrat** [comment]
@kwikius 
yeah, but it would be already a big improvement if we would change these macros.

**Comment by lucasdemarchi** [comment]
> @dgrat @lucasdemarchi Enough FUD thankyou regarding arm gcc. I suggest you substantiate your claims by testing before coming out with these daft comments

Sorry, I can't find anything I said that would trigger this behavior. If it was about std::function, I think I already explained 1) it didn't compile with the arm toolchain we had at the time I had to implement functor; I suggest you to see the log on that file and even how I couldn't include some std headers back then; 2) no constexpr constructor that I could find: it means that the tables end up in the non-initialized section in the .elf and is constructed at runtime - this may not be that relevant for Linux, but for px4 it means it can't leave that table in flash; 3) it allocates memory from heap to store the obj pointer / callback.... This is probably the case why we don't have a constexpr (that I could find whole implementing the function)
 
These 3 points were pretty much tested to be true. At the time functor was implemented. So... Sometimes we need to have some compromises on what we can or can't use from std. It may be even wrong or historical... Evolving the code is natural and if it works now it doesn't mean it's FUD or something else.

**Comment by kwikius** [comment]
OK Apologies. Let us move on. :)
At least I found the issue with the AP_Scheduler example on px4. It turns out there is a startup script on the romfs which should look for and start any imus it can find, but it appears that the version of the script used for the example doesnt ask to start the imu, so imu is not found when Ardupilor example tries to start them. I think that is the problem, but I have no idea where these startup scripts are created from and how/ where they are installed. I think there is a way too do it by putting one on the SD card ... so will try that  next time round"

**Pull request closed without merging**